### ISSUE TITLE
Presentation XML caption refactoring

### ISSUE THREAD

**Issue by opoudjis** [metanorma/mn-native-pdf#770]
This comes out of https://github.com/metanorma/isodoc/issues/617

The refactoring of captions in Presentation XML is extremely slow for me: it has taken me three weeks of effort just to get through metanorma-iso, and I will likely be refactoring isodoc as I go with other flavours. The intention is to do a release in two weeks time, and I already do not think I will make that for all of Metanorma.

For that reason, it is very important to keep all work in a separate branch until all flavours of Metanorma are done. Even more critical, this refactor is now breaking STS (https://github.com/metanorma/mnconvert/issues/418), so we may need to delay release until STS is addressed as well. And IEEE XML.

I will be updating you with each flavour as I finish it. Right now, ISO is the only flavour switched across.

This is the largest task because it impacts the most elements, but I am hoping it is much less work for you to do: it is adding semantic markup which you should be able to ignore, and moving captions and titles to a Presentation XML-specific element, which you will be using instead of /title and /name.

The updates are:

* For all blocks (requirement, permission, recommendation, formula, table, figure, note, admonition, example, sourcecode, ul, ol, dl, term, termexample, termnote), the caption to be rendered is `fmt-name`. `name` if present is the Semantic XML element, consisting of just the user-supplied caption, and it is to be ignored. `fmt-name` will contain the autonumber and the block type.
* For all clauses (foreword, introduction, acknowledgement, abstract, annex, appendix, terms, references, appendix), the title to be rendered is `fmt-title`. `title` if present is the Semantic XML element, consisting of just the user-supplied title, and it is to be ignored.  `fmt-title` will contain the autonumber and the block type.
  * The `fmt-*` tags will be a design pattern in the new Presentation XML, of Presentation XML elements being marked up alongside the source Semantic XML elements, rather than overwriting them. In the refactoring, I will be introducing a lot more of these. Either the `fmt-*` elements or their child `fmt-*` elements hyperlink to the Semantic XML tags they are derived from through their `source` attribute. 
* Every instance of `fmt-title` and `fmt-name` will be accompanied by a `fmt-xref-label` element, indicating what label to use for cross-references to this block or clause. You are to ignore this element.
* There will be a lot of `span` elements introduced inline, semantically annotating various kinds of delimiter in captions and cross-references to captions. Their `class` attribute will be `fmt-*`. Just render their contents.
* Autonumbering in captions and cross-references will be marked up with the `semx` element, which hyperlinks the autonumbering information to the element it is numbering. As with `span`, you need only render its contents.
  * The `element` attribute gives the name of the Semantic XML element that the `semx` element derives its value from. The `source` element gives the GUID of the  Semantic XML element that the `semx` element derives its value from.
* All autonumbered blocks and  clauses will have an `autonum` attribute giving the autonumber value.
* Hierarchically formed autonumbers for blocks will have multiple `semx` values, one for each block being referenced in the hierarchy. So figure 1-3 will be marked up as `<semx source={GUID for Figure 1}>1</semx> - <semx source={GUID for Subfigure 3}>3</semx>`.
  * I have not yet done this for clause numbers, and I don't think there's as compelling a reason to.
  * Container-based xrefs (e.g. "Clause 3.1, Note 4") will hyperlink their components separately.
  

Example, Presentation XML before:

````
          <clause id="widgets"  displayorder='5'>
            <title depth="1">3
              <tab/>
              Widgets</title>
            <clause id="widgets1" inline-header="true">
              <title>3.1</title>
              <note id="note1">
                <name>NOTE 1</name>
                <p id="_">These results are based on a study carried out on three different
                          types of kernel.
                        </p>
              </note>
              <note id="note2">
                <name>NOTE 2</name>
                <p id="_">These results are based on a study carried out on three different
                          types of kernel.
                        </p>
              </note>
              <p>
                <xref target="note1">Note 1</xref>
                <xref target="note2">Note 2</xref>
              </p>
            </clause>
          </clause>
````

Example, Presentation XML after (GUIDs stripped for test use):

````
            <clause id="widgets" displayorder="5">
                <title id="_">Widgets</title>
                <fmt-title depth="1">
                   <span class="fmt-caption-label">
                      <semx element="autonum" source="widgets">3</semx>
                      <span class="fmt-caption-delim">
                         <tab/>
                      </span>
                      <semx element="title" source="_">Widgets</semx>
                   </span>
                </fmt-title>
                <fmt-xref-label>
                   <span class="fmt-element-name">Clause</span>
                   <semx element="autonum" source="widgets">3</semx>
                </fmt-xref-label>
                <clause id="widgets1" inline-header="true">
                   <fmt-title depth="2">
                      <span class="fmt-caption-label">
                         <semx element="autonum" source="widgets1">3.1</semx>
                      </span>
                   </fmt-title>
                   <fmt-xref-label>
                      <semx element="autonum" source="widgets1">3.1</semx>
                   </fmt-xref-label>
                   <note id="note1" autonum="1">
                      <fmt-name>
                         <span class="fmt-caption-label">
                            <span class="fmt-element-name">NOTE</span>
                            <semx element="autonum" source="note1">1</semx>
                         </span>
                      </fmt-name>
                      <fmt-xref-label>
                         <span class="fmt-element-name">Note</span>
                         <semx element="autonum" source="note1">1</semx>
                      </fmt-xref-label>
                      <p id="_">These results are based on a study carried out on three different types of kernel.</p>
                   </note>
                   <note id="note2" autonum="2">
                      <fmt-name>
                         <span class="fmt-caption-label">
                            <span class="fmt-element-name">NOTE</span>
                            <semx element="autonum" source="note2">2</semx>
                         </span>
                      </fmt-name>
                      <fmt-xref-label>
                         <span class="fmt-element-name">Note</span>
                         <semx element="autonum" source="note2">2</semx>
                      </fmt-xref-label>
                      <p id="_">These results are based on a study carried out on three different types of kernel.</p>
                   </note>
                   <p>
                      <xref target="note1">
                         <span class="fmt-element-name">Note</span>
                         <semx element="autonum" source="note1">1</semx>
                      </xref>
                      <xref target="note2">
                         <span class="fmt-element-name">Note</span>
                         <semx element="autonum" source="note2">2</semx>
                      </xref>
                   </p>
                </clause>
           </clause>
````

As noted: ignore `title` and `name` and use `fmt-title` and `fmt-name` instead. Ignore `fmt-xref-label`. Render the contents of any `semx` and `span` elements.

This mass introduction of `semx` and `span` into markup means that we will have a lot more mixed content in our XML than we used to: all titles, block captions, and xrefs will contain tags. Given what has been happening with docidentifier containing tags in PDF, that means you will need to ensure that rendering copes with all that.

**Comment by opoudjis** [comment]
Note that work is being done in PR branch feature/presxml-autonum. This is the Gemfile.devel for metanorma-iso:

````
gem "isodoc", git: "https://github.com/metanorma/isodoc", branch: "feature/presxml-autonum"
gem "isodoc-i18n", git: "https://github.com/metanorma/isodoc-i18n", branch: "fix/markup-connectives"
gem "mn-requirements", git: "https://github.com/metanorma/mn-requirements", branch: "feature/presxml-autonum"
````

**Comment by opoudjis** [comment]
I will now be inserting any tabs between the note label and note content in presentation XML, so you won't need to:

Before, ISO:

````
<note>
  <name>NOTE</name>
````

After, ISO:

````
<note>
  <fmt-name>
         <span class="fmt-caption-label">
           <span class="fmt-element-name">NOTE</span>
         </span>
         <span class="fmt-label-delim">
            <tab/>
         </span>
  <fmt-name>
````

Before, IEEE:

````
<note>
  <name>NOTE — </name>
````

After, ISO:

````
<note>
  <fmt-name>
         <span class="fmt-caption-label">
           <span class="fmt-element-name">NOTE</span>
         </span>
         <span class="fmt-label-delim">—</span>
  <fmt-name>
````

**Comment by Intelligent2013** [comment]
I think the best way is to do these actions in the the templates `mode="update_xml_step1"` (common.xsl):

- in the elements `requirement`, `permission`, `recommendation`, `formula`, `table`, `figure`, `note`, `admonition`, `example`, `sourcecode`, `ul`, `ol`, `dl`, `term`, `termexample`, `termnote`,
OR if the element `name` `fmt-name` nearby
  - [ ] remove `name`
  - [ ] rename `fmt-name` to `name` 


- in the elements `foreword`, `introduction`, `acknowledgement`, `abstract`, `annex`, `appendix`, `terms`, `references`,
OR if the element `title` `fmt-title` nearby
   - [ ] remove `title`
   - [ ] rename `fmt-title` to `title`


- [ ] remove `fmt-xref-label`
	
- [ ] add the processing for the new elements `span`

**Comment by Intelligent2013** [comment]
I have to comment `<xsl:strip-space elements="iso:xref"/>` in common.xsl (added in https://github.com/metanorma/metanorma-iso/issues/852), because with stripped spaces inside xref the XML:
```
<xref type="inline" target="ISO20483"><span class="stdpublisher">ISO </span><span class="stddocNumber">20483</span>:<span class="stdyear">2013</span>, <span class="citeapp">Annex C</span> <span class="fmt-conn">and</span> <span class="citetbl">Table C.1</span></xref>
```
will be rendered so:
`Annex CandTable C.1`

and will review the solution for issue https://github.com/metanorma/metanorma-iso/issues/852.

**Comment by opoudjis** [comment]
Yes, that is what I was afraid of full stop that's why I noted that there's going to be much more mixed content now than before

**Comment by Intelligent2013** [comment]
ISO XSLT updated in https://github.com/metanorma/mn-native-pdf/pull/773.

@opoudjis I've generated the Presentation XML for https://github.com/metanorma/mn-samples-ieee/tree/main/sources/p987.6, and found a bug - `<semx element="autonum"` contains `—`:
```
<term id="term-input-reference-axis">
	<p>
		<strong>input reference axis</strong>: The direction of an axis. <em>Syn:</em>
		<strong>IRA</strong>. </p>
	<termnote id="_6ec31af7-e76a-f673-6493-14d481e81b77" autonum="NOTE">
		<fmt-name>
			<span class="fmt-caption-label">NOTE <semx element="autonum" source="_6ec31af7-e76a-f673-6493-14d481e81b77"/>—</span>
			<span class="fmt-label-delim">—</span>
		</fmt-name>
```

Therefore it renders as long/double dash (first from `<semx element="autonum"` , second from `<span class="fmt-label-delim">`):
![image](https://github.com/user-attachments/assets/58423d96-506b-460d-ba74-18e146da0e72)

My `Gemfile`:
```
source "https://rubygems.org"

gem "metanorma-cli"

gem "metanorma-ieee", git: "https://github.com/metanorma/metanorma-ieee", branch: "feature/presxml-autonum"
gem "isodoc", git: "https://github.com/metanorma/isodoc", branch: "feature/presxml-autonum"
gem "isodoc-i18n", git: "https://github.com/metanorma/isodoc-i18n", branch: "fix/markup-connectives"
gem "mn-requirements", git: "https://github.com/metanorma/mn-requirements", branch: "feature/presxml-autonum"

gem "sassc"
```

**Comment by opoudjis** [comment]
Thank you for finding that @Intelligent2013! But I am still not done with testing and therefore debugging metanorma-ieee; in fact, I am repeatedly going back and doing refactoring on isodoc as I find issues and redundancies downstream. With luck I may get through the rest of metanorma-ieee tonight.

**Comment by opoudjis** [comment]
metanorma-ieee done

**Comment by opoudjis** [comment]
In https://github.com/metanorma/metanorma-standoc/issues/312, I had added an xref attribute  to annex/title at your request @Intelligent2013 , as a processing hint for PDF:


> Is there a possibility to include 'Appendix' title and number/letter into a separate element or attribute? When I calculated it via XSLT it was a simple to use one algorithm to put appendix number anywhere, but in presentation XML I have to calculated it again via xslt (it's a potential source of issues due possible different algorithms in presentation XML generation and XSLT) or extract from element title, but it's language specific problem.

This information is now available as /annex/fmt-xref-label (and I will do some refactoring to also make it retrievable from  /annex/fmt-name/ ). I request that I get rid of the xref attribute.

**Comment by opoudjis** [comment]
metanorma-nist done

**Comment by Intelligent2013** [comment]
IEEE XSLT updated in https://github.com/metanorma/mn-native-pdf/pull/773.

**Comment by Intelligent2013** [comment]
ITU XSLT updated in https://github.com/metanorma/mn-native-pdf/pull/773.

**Comment by Intelligent2013** [comment]
NIST XSLT updated in https://github.com/metanorma/mn-native-pdf/pull/773 and https://github.com/metanorma/mn-native-pdf-nist/pull/82.

> This information is now available as /annex/fmt-xref-label (and I will do some refactoring to also make it retrievable from /annex/fmt-name/ ). I request that I get rid of the xref attribute.

@opoudjis FYI, I've updated XSLT to retrieve from `annex/fmt-xref-label` or from `annex/fmt-title/@xref` (if exists). So, don't need to do it urgently.

**Comment by opoudjis** [comment]
metanorma-ogc done

**Comment by opoudjis** [comment]
> NIST XSLT updated in #773 and [metanorma/mn-native-pdf-nist#82](https://github.com/metanorma/mn-native-pdf-nist/pull/82).
> 
> > This information is now available as /annex/fmt-xref-label (and I will do some refactoring to also make it retrievable from /annex/fmt-name/ ). I request that I get rid of the xref attribute.
> 
> @opoudjis FYI, I've updated XSLT to retrieve from `annex/fmt-xref-label` or from `annex/fmt-title/@xref` (if exists). So, don't need to do it urgently.

It's now `annex/fmt-title/span[@class = 'fmt-caption-label']`, and `@xref` is no longer generated.

**Comment by opoudjis** [comment]
metanorma-csa blank document test is aborting:

````
       "fo:block-container" is missing child elements. Required content model: marker* (%block;)+ (See position 170:81)
       org.apache.fop.fo.ValidationException: "fo:block-container" is missing child elements. Required content model: marker* (%block;)+ (See position 170:81)
       javax.xml.transform.TransformerException: org.apache.fop.fo.ValidationException: "fo:block-container" is missing child elements. Required content model: marker* (%block;)+ (See position 170:81)
````

**Comment by opoudjis** [comment]
metanorma-csa done

**Comment by opoudjis** [comment]
metanorma-cc done

**Comment by opoudjis** [comment]
metanorma-iho done

**Comment by opoudjis** [comment]
metanorma-ribose done

**Comment by Intelligent2013** [comment]
OGC XSLT updated in https://github.com/metanorma/mn-native-pdf/pull/773.

**Comment by Intelligent2013** [comment]
> metanorma-csa blank document test is aborting:

@opoudjis how can I generate the Presentation XML for the blank document?
For adoc:
```adoc
= Information Security Management through Reflexive Security -- Six Pillars in the Integration of Security, Development and Operations
:doctype: standard
:docnumber: csa-01
:copyright-year: 2019
:mn-document-class: csa
:status: published
:language: en
:docfile: document.adoc
:imagesdir: images
```
the PDF generated ok.

**Comment by Intelligent2013** [comment]
CSA XSLT updated in https://github.com/metanorma/mn-native-pdf/pull/773.

**Comment by opoudjis** [comment]
> > metanorma-csa blank document test is aborting:
> 
> @opoudjis how can I generate the Presentation XML for the blank document? For adoc:
> 
> ```
> = Information Security Management through Reflexive Security -- Six Pillars in the Integration of Security, Development and Operations
> :doctype: standard
> :docnumber: csa-01
> :copyright-year: 2019
> :mn-document-class: csa
> :status: published
> :language: en
> :docfile: document.adoc
> :imagesdir: images
> ```
> 
> the PDF generated ok.

I'm still getting it aborting. The test (which is long-standing) is just:

````
= Document title
Author
:docfile: test.adoc
:novalid:
````

**Comment by Intelligent2013** [comment]
> I'm still getting it aborting. The test (which is long-standing) is just

The reason of error in the missing element `clause/title`, there is just `clause/fmt-title`:
```xml
<clause type="toc" id="_6b43a7a2-22e9-433d-9079-a1e5e401f648" displayorder="1">
	<fmt-title depth="1">Table of contents</fmt-title>
</clause>
```

Wrong XSL-FO with missing `<fo:block` inside `<fo:block-container`:
```xml
<fo:block-container color="rgb(7, 72, 156)"
                                font-size="12pt"
                                line-height="170%">Table of contents</fo:block-container>
```

The correct XSL-FO is:
```
<fo:block-container color="rgb(7, 72, 156)"
                                font-size="12pt"
                                line-height="170%">
                <fo:block color="black"
                          font-size="26pt"
                          margin-bottom="30pt"
                          margin-top="2pt"
                          role="H1">Table of contents</fo:block>
       </fo:block-container>
```

I'll update the XSLT.

**Comment by Intelligent2013** [comment]
CSA XSLT updated in `main` branch in https://github.com/metanorma/metanorma-csa.

**Comment by Intelligent2013** [comment]
CSD XSLT updated in https://github.com/metanorma/mn-native-pdf/pull/773.

**Comment by Intelligent2013** [comment]
> metanorma-iho done

@opoudjis  I can't generate the Presentation XML for the document https://github.com/metanorma/mn-samples-iho/tree/main/sources/s102
`bundle exec metanorma -t iho -x presentation document.adoc`:
```
[relaton] INFO: (ISO/TS 19103) not found in cache, if you wish to ignore cache please use `no-cache` option.
[relaton] INFO: (ISO 19129) not found in cache, if you wish to ignore cache please use `no-cache` option.
AsciiDoc Input: Missing workgroup attribute for document
Style: (ID sec-overview): Hanging paragraph in clause
Style: (ID _terms_definitions_and_abbreviations): Hanging paragraph in clause
Style: (ID subsec-application-schema): Hanging paragraph in clause
Style: (ID subsec-tiling-scheme-partitioning): Hanging paragraph in clause
Style: (ID _feature_types): Hanging paragraph in clause
Style: (ID sec-data-quality): Hanging paragraph in clause
Style: (ID _generation_and_display_of_gridded_bathymetry): Hanging paragraph in clause
Style: (ID subsec-product-structure): Hanging paragraph in clause
Style: (ID subsec-dataset-size): Hanging paragraph in clause
Style: (ID _data_integrity_and_encryption): Hanging paragraph in clause
Style: (ID _structure_metadata): Hanging paragraph in clause
Style: (ID subsec-s102_exchangecatalogue): Hanging paragraph in clause
Style: (ID _s102_datasetdiscoverymetadata): Hanging paragraph in clause
Style: (ID _s102_cataloguemetadata): Hanging paragraph in clause
Style: (ID _modern_high_resolution_hydrographic_multibeam_sonars): Hanging paragraph in clause
Style: (ID _7bcb1bed-e2f3-a93f-108a-7318c5916426): Table should have title
C:/tools/ruby31/lib/ruby/gems/3.1.0/bundler/gems/isodoc-bf95855d8879/lib/isodoc/presentation_function/autonum.rb:85:in `[]'
C:/tools/ruby31/lib/ruby/gems/3.1.0/bundler/gems/isodoc-bf95855d8879/lib/isodoc/presentation_function/autonum.rb:85:in `fmt_caption2'
C:/tools/ruby31/lib/ruby/gems/3.1.0/bundler/gems/isodoc-bf95855d8879/lib/isodoc/presentation_function/autonum.rb:69:in `fmt_caption'
C:/tools/ruby31/lib/ruby/gems/3.1.0/bundler/gems/isodoc-bf95855d8879/lib/isodoc/presentation_function/autonum.rb:10:in `prefix_name'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-generic-2.7.2/lib/isodoc/generic/presentation_xml_convert.rb:13:in `annex1'
C:/tools/ruby31/lib/ruby/gems/3.1.0/bundler/gems/isodoc-bf95855d8879/lib/isodoc/presentation_function/section.rb:44:in `block in annex'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/nokogiri-1.16.7-x64-mingw-ucrt/lib/nokogiri/xml/node_set.rb:235:in `block in each'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/nokogiri-1.16.7-x64-mingw-ucrt/lib/nokogiri/xml/node_set.rb:234:in `upto'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/nokogiri-1.16.7-x64-mingw-ucrt/lib/nokogiri/xml/node_set.rb:234:in `each'
C:/tools/ruby31/lib/ruby/gems/3.1.0/bundler/gems/isodoc-bf95855d8879/lib/isodoc/presentation_function/section.rb:42:in `annex'
C:/tools/ruby31/lib/ruby/gems/3.1.0/bundler/gems/isodoc-bf95855d8879/lib/isodoc/presentation_xml_convert.rb:62:in `section'
C:/tools/ruby31/lib/ruby/gems/3.1.0/bundler/gems/isodoc-bf95855d8879/lib/isodoc/presentation_xml_convert.rb:45:in `conversions'
C:/tools/ruby31/lib/ruby/gems/3.1.0/bundler/gems/isodoc-bf95855d8879/lib/isodoc/presentation_xml_convert.rb:28:in `convert1'
C:/tools/ruby31/lib/ruby/gems/3.1.0/bundler/gems/isodoc-bf95855d8879/lib/isodoc/convert.rb:147:in `convert'
C:/tools/ruby31/lib/ruby/gems/3.1.0/bundler/gems/metanorma-iho-5da79828674a/lib/metanorma/iho/processor.rb:32:in `output'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-2.0.7/lib/metanorma/compile/compile.rb:193:in `process_output_unthreaded'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-2.0.7/lib/metanorma/compile/compile.rb:179:in `process_exts1'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-2.0.7/lib/metanorma/compile/compile.rb:152:in `process_ext'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-2.0.7/lib/metanorma/compile/compile.rb:118:in `process_exts'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-2.0.7/lib/metanorma/compile/compile.rb:35:in `compile'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-cli-1.11.1/lib/metanorma/cli/compiler.rb:45:in `compile_file'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-cli-1.11.1/lib/metanorma/cli/compiler.rb:30:in `compile'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-cli-1.11.1/lib/metanorma/cli/compiler.rb:35:in `compile'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-cli-1.11.1/lib/metanorma/cli/command.rb:272:in `compile_document'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-cli-1.11.1/lib/metanorma/cli/command.rb:62:in `block in compile'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-cli-1.11.1/lib/metanorma/cli/command.rb:62:in `each'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-cli-1.11.1/lib/metanorma/cli/command.rb:62:in `compile'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/thor-1.3.2/lib/thor/command.rb:28:in `run'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/thor-hollaback-0.2.1/lib/thor/hollaback.rb:68:in `run'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/thor-1.3.2/lib/thor/invocation.rb:127:in `invoke_command'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/thor-1.3.2/lib/thor.rb:538:in `dispatch'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/thor-1.3.2/lib/thor/base.rb:584:in `start'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-cli-1.11.1/lib/metanorma/cli.rb:34:in `start'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-cli-1.11.1/exe/metanorma:25:in `block in <top (required)>'
C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-cli-1.11.1/exe/metanorma:43:in `<top (required)>'
C:/tools/ruby31/bin/metanorma:25:in `load'
C:/tools/ruby31/bin/metanorma:25:in `<main>'
no implicit conversion of Symbol into Integer
```

`Gemfile`:
```
gem "metanorma-cli"

gem "metanorma-iho", git: "https://github.com/metanorma/metanorma-iho", branch: "feature/presxml-autonum"
gem "isodoc", git: "https://github.com/metanorma/isodoc", branch: "feature/presxml-autonum"
gem "isodoc-i18n", git: "https://github.com/metanorma/isodoc-i18n", branch: "fix/markup-connectives"
gem "mn-requirements", git: "https://github.com/metanorma/mn-requirements", branch: "feature/presxml-autonum"

gem "debug"
```

With `Gemfile`:
```
gem "metanorma-cli"

gem "debug"
```
the process ends successfully.

**Comment by Intelligent2013** [comment]
With commented annexes the Presentation XML generates ok:
```
////

include::sections/aa-data_class_enc_.adoc[]

include::sections/ac-norm_impl_guidance.adoc[]

include::sections/ad-feature_catalogue.adoc[]

include::sections/ae-portrayal-catalogue.adoc[]

include::sections/af-dataset_size_production.adoc[]

include::sections/ag-gridding_methods.adoc[]

include::sections/ah-multi-res_gridding.adoc[]

include::sections/ai-gridding_full.adoc[]

////
```

**Comment by Intelligent2013** [comment]
Ribose XSLT updated in https://github.com/metanorma/mn-native-pdf/pull/773.

**Comment by ronaldtse** [comment]
> With commented annexes the Presentation XML generates ok:

This means the refactor in Metanorma-IHO has not been fully tested and needs to be fixed.

**Comment by opoudjis** [comment]
> > With commented annexes the Presentation XML generates ok:
> 
> This means the refactor in Metanorma-IHO has not been fully tested and needs to be fixed.

And when I am done, I will run the thing locally through all the samples directories. I am not at that stage yet.

**Comment by opoudjis** [comment]
I cannot replicate your issue: I am generating S-102 fine, although PDF is crashing.

Gemfile reads:

````

````
source "https://rubygems.org"

gem "metanorma-cli"
gem "sassc"
gem "debug"

gem "isodoc", git: "https://github.com/metanorma/isodoc", branch: "feature/presxml-autonum"
gem "metanorma-generic", git: "https://github.com/metanorma/metanorma-generic", branch: "feature/presxml-autonum"
gem "metanorma-iho", git: "https://github.com/metanorma/metanorma-iho", branch: "feature/presxml-autonum"
gem "isodoc-i18n", git: "https://github.com/metanorma/isodoc-i18n", branch: "fix/markup-connectives"
gem "mn-requirements", git: "https://github.com/metanorma/mn-requirements", branch: "feature/presxml-autonum"
````

Annexes are uncommented:

````
[[annex-data-classification-and-encoding-guide]]
[appendix]
== Data Classification and Encoding Guide

=== Features

==== BathymetryCoverage
````

I can confirm that your trace reflects the latest isodoc, and is on the line:

````
        label.blank? or
          d = "<span class='fmt-caption-delim'>#{delims[:caption]}</span>"
````

prefix_name() always should be passed a `delims` hash, for the caption delimiter (between autonum and title) and the label delimiter (after the title). I'm going to defensively set it to {} if it is nil, just in case.

**Comment by opoudjis** [comment]
Getting crash for empty document in JIS, may be the same issue:

````
"fo:flow" is missing child elements. Required content model: marker* (%block;)+ (See position 20:2184)
org.apache.fop.fo.ValidationException: "fo:flow" is missing child elements. Required content model: marker* (%block;)+ (See position 20:2184)
javax.xml.transform.TransformerException: org.apache.fop.fo.ValidationException: "fo:flow" is missing child elements. Required content model: marker* (%block;)+ (See position 20:2184)
	at org.apache.xalan.transformer.TransformerIdentityImpl.transform(TransformerIdentityImpl.java:536)
	at org.metanorma.fop.PDFGenerator.runFOP(PDFGenerator.java:663)
	at org.metanorma.fop.PDFGenerator.convertmn2pdf(PDFGenerator.java:489)
	at org.metanorma.fop.PDFGenerator.process(PDFGenerator.java:320)
	at org.metanorma.fop.mn2pdf.main(mn2pdf.java:338)
````

`````
      = Document title
      Author
      :docfile: test.adoc
      :novalid:
      :docnumber: 0

      == Clause
`````

**Comment by Intelligent2013** [comment]
> I cannot replicate your issue: I am generating S-102 fine, although PDF is crashing.

I've added 
```
gem "metanorma-generic", git: "https://github.com/metanorma/metanorma-generic", branch: "feature/presxml-autonum"
```
and XML generated now. Thanks.

**Comment by Intelligent2013** [comment]
> metanorma-iho done

@opoudjis in the Presentation XML, `table` and `figure` `fmt-name/span class="fmt-caption-label"/semx element="autonum"` has the redundant prefix `Annex `:

```xml
<table id="_685ade28-45c8-a903-15fd-92cbcfa86f18" autonum="Annex A-2">
	<name id="_ff92731a-df8d-4213-8697-1a069fde4b83">BathymetryCoverage feature attribute parameters</name>
	<fmt-name>
		<span class="fmt-caption-label">
			<span class="fmt-element-name">Table</span>
			<semx element="autonum" source="annex-data-classification-and-encoding-guide">Annex A</semx>
			<span class="fmt-autonum-delim">-</span>
			<semx element="autonum" source="_685ade28-45c8-a903-15fd-92cbcfa86f18">2</semx>
		</span>
		<span class="fmt-caption-delim"> — </span>
		<semx element="name" source="_ff92731a-df8d-4213-8697-1a069fde4b83">BathymetryCoverage feature attribute parameters</semx>
	</fmt-name>

...

<figure id="fig-informative-grid-extents-for-a-10mb-uncompressed-dataset" autonum="Annex C-1">
	<name id="_1c937634-b867-4997-9c99-26dc2a967c6c">Informative grid extents for a 10 MB Uncompressed Dataset</name>
	<fmt-name>
		<span class="fmt-caption-label">
			<span class="fmt-element-name">Figure</span>
			<semx element="autonum" source="annex-s102-dataset-size-and-production">Annex C</semx>
			<span class="fmt-autonum-delim">-</span>
			<semx element="autonum" source="fig-informative-grid-extents-for-a-10mb-uncompressed-dataset">1</semx>
		</span>
		<span class="fmt-caption-delim"> — </span>
		<semx element="name" source="_1c937634-b867-4997-9c99-26dc2a967c6c">Informative grid extents for a 10 MB Uncompressed Dataset</semx>
	</fmt-name>	
...
```
PDF rendering sample:
![image](https://github.com/user-attachments/assets/a729dbbf-3518-463f-b027-62873a888ec5)

**Comment by Intelligent2013** [comment]
IHO XSLT updated in https://github.com/metanorma/mn-native-pdf/pull/773.

**Comment by Intelligent2013** [comment]
> Getting crash for empty document in JIS, may be the same issue:

@opoudjis is this entry in `Gemfile` correct?
```
gem "metanorma-jis", git: "https://github.com/metanorma/metanorma-jis", branch: "feature/presxml-autonum"
```
Then could you add the branch `feature/presxml-autonum`, please?

**Comment by opoudjis** [comment]
i am still working on it, and have been for the past 8 HOURS. It is not ready yet.

**Comment by opoudjis** [comment]
metanorma-jis ready

**Comment by opoudjis** [comment]
> > metanorma-iho done
> 
> @opoudjis in the Presentation XML, `table` and `figure` `fmt-name/span class="fmt-caption-label"/semx element="autonum"` has the redundant prefix `Annex `:

That was deliberate: Appendixes have arabic numbering, and Figure 1-4 would ambiguous between Clause 1 and Appendix 1: https://github.com/metanorma/metanorma-iho/issues/77 . But I will remove that behaviour for Annexes.

**Comment by opoudjis** [comment]
metanorma-iec done

**Comment by Intelligent2013** [comment]
> metanorma-jis ready

@opoudjis issues found
- the characters `.` (`fullwidth full stop`, U+FF0E) instead of  `・` (`katakana middle dot`, U+30FB) between digits in the clause numbers:
New presentation XML (`<span class="fmt-autonum-delim">．</span>`):
```xml
<terms id="sec_3.1" obligation="normative">
	<title id="_38fc86da-2d2c-4df7-9dbb-f7d967fe79c6">規格の種類</title>
	<fmt-title depth="2">
		<span class="fmt-caption-label">
			<semx element="autonum" source="_terms_and_definitions">三</semx>
			<span class="fmt-autonum-delim">．</span>
			<semx element="autonum" source="sec_3.1">一</semx>
		</span>
```

Old Presentation XML:
```xml
<title depth="2">三・一<tab/>規格の種類</title>
```

See https://github.com/metanorma/metanorma-jis/issues/228#issuecomment-2432791000.

-  wrong numbering in `annex/clause/clause`
New Presentation XML:
```xml
<clause id="___x69cb___x6210___x8981___x7d20___x306b___x3064___x3044___x3066_" inline-header="false" obligation="informative">
	<title id="_40c6baa6-2306-46d0-8caa-7c62a739fd55">構成要素について</title>
	<fmt-title depth="2">
		<span class="fmt-caption-label">
			<semx element="autonum" source="___x69cb___x6210___x8981___x7d20___x306b___x3064___x3044___x3066_">五</semx>
		</span>
		
		...
	
	<clause id="___x5168___x4f53_" inline-header="false" obligation="informative">
		<title id="_60a256b9-a24f-414a-9366-9b9783e8b259">全体</title>
		<fmt-title depth="3">
			<span class="fmt-caption-label">
				<semx element="autonum" source="___x69cb___x6210___x8981___x7d20___x306b___x3064___x3044___x3066_">五</semx>
				<span class="fmt-autonum-delim">．</span>
				<semx element="autonum" source="___x5168___x4f53_">五・一</semx>
			</span>
			<span class="fmt-caption-delim">
				<tab/>
			</span>
```
PDF rendering:
![image](https://github.com/user-attachments/assets/63edc77b-37ec-44e3-aac0-7713f068f2f6)

Old Presentation XML:
```xml
<clause id="___x69cb___x6210___x8981___x7d20___x306b___x3064___x3044___x3066_" inline-header="false" obligation="informative">
	<title depth="2">五<tab/>構成要素について</title>
	<clause id="___x5168___x4f53_" inline-header="false" obligation="informative">
		<title depth="3">五・一<tab/>全体</title>
```

PDF rendering:
![image](https://github.com/user-attachments/assets/7e3702c3-738b-42e9-bf08-1fe12878739c)


There are 1735 differences between old and new XSL-FO. May be there are another issues, so I can find another only after these two fixing.

**Comment by opoudjis** [comment]
Will investigate before continuing. Remaining flavours are BSI, BIPM, Plateau.

**Comment by opoudjis** [comment]
Sorry it's taken so long for me to investigate this, but after 14 hours straight yesterday, I was exhausted. 

I don't see it from the screenshot, but is your second bug that A.1.1 is ending up as A.一.1 instead of A・一・一? I'm fixing that bug, but it doesn't correspond to your screenshot.

**Comment by opoudjis** [comment]
@Intelligent2013 Done, you'll need to pull both metanorma-iso and metanorma-iso.

**Comment by opoudjis** [comment]
I'm still working on this, but the Semantic XML elements floating-title and section-title are about to show up in Presentation XML (the latter specific to BSI). You are to ignore them completely in rendering: their content is replicated in paragraphs of type floating-title and section-title for Presentation XML, as has already been the case.

**Comment by Intelligent2013** [comment]
> I don't see it from the screenshot, but is your second bug that A.1.1 is ending up as A.一.1 instead of A・一・一? I'm fixing that bug, but it doesn't correspond to your screenshot.

The _second_ level clause in annex has _3_ numbers separated by dot, i.e. `五．五・一` instead of `五・一`.
I.e. `五．` is redundant.

**Comment by opoudjis** [comment]
> > I don't see it from the screenshot, but is your second bug that A.1.1 is ending up as A.一.1 instead of A・一・一? I'm fixing that bug, but it doesn't correspond to your screenshot.
> 
> The _second_ level clause in annex has _3_ numbers separated by dot, i.e. `五．五・一` instead of `五・一`. I.e. `五．` is redundant.

Right, I see (though the separate issue I found was also a bug). That's a distinct bug, in commentary annex subclauses, which are treated separately. It's 3 am, so I'll debug this tomorrow.

**Comment by Intelligent2013** [comment]
> Done, you'll need to pull both metanorma-iso and metanorma-iho.

Confirmed. Thanks!

**Comment by Intelligent2013** [comment]
IEC XSLT updated in https://github.com/metanorma/mn-native-pdf/pull/773.

**Comment by opoudjis** [comment]
> > > I don't see it from the screenshot, but is your second bug that A.1.1 is ending up as A.一.1 instead of A・一・一? I'm fixing that bug, but it doesn't correspond to your screenshot.
> > 
> > 
> > The _second_ level clause in annex has _3_ numbers separated by dot, i.e. `五．五・一` instead of `五・一`. I.e. `五．` is redundant.
> 
> Right, I see (though the separate issue I found was also a bug). That's a distinct bug, in commentary annex subclauses, which are treated separately. It's 3 am, so I'll debug this tomorrow.

Fixed.

**Comment by opoudjis** [comment]
I'm sorry about all these coming through, the XML is a lot noisier now, so it is making it easier to miss things in unit testing when I've been starting at a screen for 14 hours...

**Comment by Intelligent2013** [comment]
> Fixed.

Confirmed. Thanks!
JIS XSLT updated in https://github.com/metanorma/mn-native-pdf/pull/773.

> I'm sorry about all these bugs coming through, the XML is a lot noisier now, so it is making it easier to miss things in unit testing when I've been starting at a screen for 14 hours...

No problem at all.

**Comment by opoudjis** [comment]
metanorma-bipm done

**Comment by opoudjis** [comment]
metanorma-plateau done

**Comment by Intelligent2013** [comment]
> metanorma-bsi done

@opoudjis issues:
- `li/@label` is empty:
```xml
<ol id="_69327256-15e9-228e-ece9-e718122a623a" type="alphabet">
	<li label="">
```
- there is a space between `semx`:
```xml
<p id="_d48ff641-06ae-d425-8fff-8576944732e0">Hardwood block (see <xref target="fig2b"><span class="citefig"><span class="fmt-element-name">Figure</span> <semx element="autonum" source="fig2">2</semx> <semx element="autonum" source="fig2b">b</semx><span class="fmt-autonum-delim">)</span></span></xref>)</p>
```
PDF renders with space (the source adoc https://github.com/metanorma/mn-samples-bsi/tree/main/sources/bs-1234-5-2022):
![image](https://github.com/user-attachments/assets/76ea814e-4eb3-418b-9316-ac108114634c)

Expected result:
![image](https://github.com/user-attachments/assets/a7b3e2c8-437c-4e4a-8be4-e6307b50ead8)

**Comment by Intelligent2013** [comment]
BSI XSLT updated in https://github.com/metanorma/mn-native-pdf/pull/773.

**Comment by Intelligent2013** [comment]
> metanorma-bipm done

@opoudjis  issue found - in the Index there are both references to the `title/bookmark` and `fmt-title/bookmark`:

```xml
<clause id="_appendix_3_the_base_unitsbase_units_of_the_si" obligation="normative" unnumbered="true">
	<title id="_18cc76a6-96da-4886-9d62-248e129a14fb">Appendix 3. The base units<bookmark id="_a17d01cc-8089-4eb9-88be-33ff27fad132"/> of the SI</title>
	<fmt-title depth="5">
		<semx element="title" source="_18cc76a6-96da-4886-9d62-248e129a14fb">Appendix 3. The base units<bookmark id="_a3cbeb24-fde5-40cc-bd17-ea2081c8bcd4"/> of the SI</semx>
	</fmt-title>
```	

```xml
<indexsect id="_17dde3a6-ae02-4a2c-b121-c963cdd0f029" displayorder="15">
...
<li>base unit(s),  ...

		<xref target="_a17d01cc-8089-4eb9-88be-33ff27fad132" pagenumber="true">"<semx element="title" source="_appendix_3_the_base_unitsbase_units_of_the_si">Appendix 3. The base unitsbase unit(s) of the SI</semx>"</xref>, 
		
		<xref target="_a3cbeb24-fde5-40cc-bd17-ea2081c8bcd4" pagenumber="true">"<semx element="title" source="_appendix_3_the_base_unitsbase_units_of_the_si">Appendix 3. The base unitsbase unit(s) of the SI</semx>"</xref>,
...
</li>
```

As the element `title` ignored, the PDF renders with spaces for missing refs:
![image](https://github.com/user-attachments/assets/57c9ec7d-0f28-4873-a1d8-8b1c2b322003)

And in PDF log there are messages:
```
Page 110: Unresolved ID reference "_a17d01cc-8089-4eb9-88be-33ff27fad132" found.
Page 110: Unresolved ID reference "_a70f554b-425c-4510-ac42-977e33286cbf" found.
Page 110: Unresolved ID reference "_8d8d5193-6bb7-4017-bb64-5e2756145d00" found.
....
```

**Comment by opoudjis** [comment]
> in the Index there are both references to the title/bookmark and fmt-title/bookmark:

That's an issue I knew was coming, but I hoped had not spread yet—that IDs are being replicated between title and fmt-title. I am dodging it in floating-title, by reassigning `floating-title/*[@id]` to `@original-id`; I need to generalise that.

**Comment by opoudjis** [comment]
In addition, I can't leave `index` items in Semantic XML elements, they need to be moved to the Presentation XML elements, globally.

**Comment by opoudjis** [comment]
Fixed, try it out now. The `<index>` is removed from `title`, it is only left in `fmt-title`.

**Comment by Intelligent2013** [comment]
Plateau XSLT updated in https://github.com/metanorma/mn-native-pdf/pull/773.

Now, I'll test the BIPM.

**Comment by Intelligent2013** [comment]
> Fixed, try it out now. The `<index>` is removed from `title`, it is only left in `fmt-title`.

@opoudjis  still there are references between `index//xref` and `title/bookmark`.
The previous PDF generation log contains 167 errors `Unresolved ID reference "..." found.`
Now, 88 errors. 
For instance, `_1d07cd96-a110-4d66-9d02-cb7fc1eede83`:
```xml
<clause id="cgpm12th1964r7" unnumbered="true" obligation="normative">
<title type="quoted"><blacksquare/><strong>Curie</strong> (<link target="https://www.bipm.org/en/committees/cg/cgpm/12-1964/resolution-7">CGPM RES 7 (1964, E)</link>)<bookmark id="_1d07cd96-a110-4d66-9d02-cb7fc1eede83"/></title><fmt-title type="quoted" depth="3"><blacksquare/><strong>Curie</strong> (<link target="https://www.bipm.org/en/committees/cg/cgpm/12-1964/resolution-7">CGPM RES 7 (1964, E)</link>)<bookmark id="_f6184b8a-e919-4531-be6f-28712e0ee20f"/></fmt-title>
```
```xml
<indexsect 
<xref target="_1d07cd96-a110-4d66-9d02-cb7fc1eede83" pagenumber="true">"<semx element="title" source="cgpm12th1964r7">[cgpm12th1964r7]</semx>"</xref>
```

**Comment by opoudjis** [comment]
Hm, ok. This is all BIPM Brochure, I just was reluctant to compile the whole thing, but clearly I'll need to...

**Comment by opoudjis** [comment]
I was catching all title/fmt-title duplications, but the one specific to BIPM with quoted titles. I am now testing the Brochure outputs to ensure no further unresolved ids (those are xrefs pointing to title or name, which are now excluded from rendering, instead of their duplicates in fmt-title and fmt-name.)

**Comment by opoudjis** [comment]
So compiling the English-language BIPM brochure, there are 1327 index entries. Of these, 88 are under title, and thus will go missing on rendering, since rendering uses fmt-title instead (which duplicates those titles); that is what you also found, and they should be the quoted titles I had missed last time—so we've clearly been looking at the same document.

Recompiling with the bug fix, the ancestors of index links are now

````
{"clause" => 503, 
"table" => 148, 
"xref" => 22, 
"note" => 26, 
"li" => 337, 
"fmt-title" => 167, 
"dl" => 26, 
"fmt-name" => 6, 
"dt" => 4}
````

There were a couple of surprises in there (there are index entries pointing to footnotes and to cross-reference text), but i think this means we are now good to go.

**Comment by Intelligent2013** [comment]
BIPM and JCGM XSLT updated in https://github.com/metanorma/mn-native-pdf/pull/773.

@opoudjis I've just found the double `::` in termnote name when I've tested JCGM document. The double `::` there is also in old Presentation XML.

Old Presentation XML:
```xml
<termnote id="_49888e25-31af-6205-b3e3-269d2cc55cfb"><name>Note 1 to entry:: </name>
```
New Presentation XML:
```xml
<fmt-name><span class="fmt-caption-label">Note <semx element="autonum" source="_49888e25-31af-6205-b3e3-269d2cc55cfb">1</semx> to entry:</span><span class="fmt-label-delim">: </span></fmt-name>
```

**Comment by Intelligent2013** [comment]
So, at this moment all XSLT updated for new Presentation XML in:
- https://github.com/metanorma/mn-native-pdf/pull/773
- https://github.com/metanorma/mn-native-pdf-nist/pull/82
- https://github.com/metanorma/mn-native-pdf-bsi/pull/86

Found issues in:
- metanorma-bsi (https://github.com/metanorma/mn-native-pdf/issues/770#issuecomment-2515205983):
   - `li/@label` is empty:
   - there is a space between `semx`
- metanorma-bipm:
  - reference from `index` to `title/bookmark`
  - double `::` in termnote name

**Comment by opoudjis** [comment]
BSI fixed. References in BIPM from index to title/bookmark should all have been fixed yesterday, I found no remaining instances. Removed redundant colon in BIPM termnotes.

**Comment by opoudjis** [comment]
Please confirm all is ready, intend to release Monday.

**Comment by Intelligent2013** [comment]
> BSI fixed,

Confirmed. Thank you!

**Comment by Intelligent2013** [comment]
> References in BIPM from index to title/bookmark should all have been fixed yesterday, I found no remaining instances. Removed redundant colon in BIPM termnotes.

Confirmed.

I'll merge the XSLT PRs today.

**Comment by Intelligent2013** [comment]
XSLTs merged into the `main` branch in `metanorma-...` repositories.
